### ISSUE TITLE
Allow awaiting tasks in async CEs without Async.AwaitTask

### ISSUE THREAD

**Issue by cmeeren** [fsharp/fslang-suggestions#1142]
I propose we allow `let!`/`do!`/`match!`/`return!` in async CEs with `Task` values.

Additionally, if possible (if there are no hidden gotchas and it's easy to implement), we could consider doing the same for `ValueTask` and/or other awaitables (anything that has a suitable `GetAwaiter` method?), though `Task` is the most important one.

The existing way of approaching this problem in F# is to use `Async.AwaitTask` (and additionally, for `ValueTask`, calling `AsTask()` first).

## Pros and Cons 

The advantages of making this adjustment to F# are:
* Reducing unnecessary noise when interoping with the sizable amount of code (in BCL and elsewhere) that uses `Task` (I say _unnecessary_ because AFAIK there is only one good way to await `Task` values, namely `Async.AwaitTask`, so it only adds noise)
* Consistency with `task` CE, which allows awaiting `Async`

The disadvantages of making this adjustment to F# are: None that I can think of.

## Extra information

Estimated cost (XS, S, M, L, XL, XXL): XS (Just add a suitable `Bind` overload to the async CE)

## Affidavit (please submit!)

Please tick this by placing a cross in the box:
* [x] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [x] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [x] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [x] This is not a breaking change to the F# language design
* [ ] I or my company would be willing to help implement and/or test this

## For Readers

If you would like to see this issue implemented, please click the :+1: emoji on this issue. These counts are used to generally order the suggestions by engagement.

**Comment by baronfel** [comment]
One personal hurdle I've had with this easy bridging mechanism is that it makes it _very_ easy to lose the CancellationToken for the async in the translation. The explicit call serves as a reminder to me to check if my Task-generating function accepts a CancellationToken, and if so extract it from the `async` context before creating the Task.

**Comment by jwosty** [comment]
> One personal hurdle I've had with this easy bridging mechanism is that it makes it very easy to lose the CancellationToken for the async in the translation

Yeah, it would be really great to solve this in some way. Would be really nice if the compiler could emit a warning when a) you're inside an async context and b) you call a task-returning method/function with an optional cancellation token parameter and c) you haven't supplied said optional token. But that would be another language suggestion.

**Comment by baronfel** [comment]
> that would be another language suggestion.

Yeah, and in the general case that would be something better served by an Analyzer IMO. Keep the compiler core itself focused, and push best-practices stuff outside the core.

**Comment by cmeeren** [comment]
> The explicit call serves as a reminder to me to check if my Task-generating function accepts a CancellationToken, and if so extract it from the `async` context before creating the Task.

Sure, though if its only real utility is acting as a reminder for checking something elsewhere (in your task-generating function), then I'd argue it's not worth keeping. As you say above, an analyzer or similar is better for that.

**Comment by jwosty** [comment]
> > The explicit call serves as a reminder to me to check if my Task-generating function accepts a CancellationToken, and if so extract it from the `async` context before creating the Task.
> 
> Sure, though if its only real utility is acting as a reminder for checking something elsewhere (in your task-generating function), then I'd argue it's not worth keeping. As you say above, an analyzer or similar is better for that.

I was referring to something more like the following:

```fsharp
// given:
//type File =
//    static member ReadAllTextAsync: path: string *
//                                    ?cancellationToken: Threading.CancellationToken
//                                    -> Task<string>

let doWork file = async {
    // Could generate warning to pass in CancellationToken here, since it is statically knowable here
    let! text = Async.AwaitTask (File.ReadAllTextAsync file)
    printfn "file contents: %A" text
}
```

And if we used a function instead of ReadAllText which doesn't take a cancellation token, it wouldn't emit said warning. So it would be a sure-fire "you forgot to pass the cancellation token, dummy." However as has already been said one can make the argument that it should be in an analyzer, not the compiler

**Comment by edgarfgp** [comment]
This took me sometime to figure out and would be awesome to have a more friendly approach

**Comment by jamil7** [comment]
> Estimated cost (XS, S, M, L, XL, XXL): XS (Just add a suitable Bind overload to the async CE)

I tried doing this a while ago, and found that overloading `_.Bind()` didn't work so well when I was mixing Asyncs and Tasks in the CE. It broke type inference quickly, and I had to annotate all the variables in the CE manually, which was annoying. 

So I used the hide and seek champion `_.Source()` instead (page 70 [here](https://fsharp.org/specs/language-spec/4.1/FSharpSpec-4.1-latest.pdf)). I'm not sure if it was intended to be used like this, but [here](https://github.com/jamil7/FSharp.Prelude/blob/2db50bdb3674e3a95f72509f9f1ce910a6916078/src/Prelude/Extensions/Async.fs#L126) is what I ended up with, which works fine in my personal experience when it comes to type inference.
